src/cliffordsynthesis/encoding/PhaseCorrectionEncoder.cpp

```diff
@@ -0,0 +1,148 @@
+    for (const auto& pauli : PAULIS) {
+      const std::string gName =
+          "g_" + std::to_string(i) + "_" + toString(pauli);
+      TRACE() << "Creating pauli variable " << gName;
+      paulis[i].emplace_back(lb->makeVariable(gName));
+    }
+  }
+
+  const std::bitset<64> init = uncorrected.getBVFrom(2 * N);
+  initialPhase               = vectorFromBitset(init);
+  const std::bitset<64> tar  = target.getBVFrom(2 * N);
+  targetPhase                = vectorFromBitset(tar);
+}
+
+void PhaseCorrectionEncoder::encodePauliConstraints() {
+  DEBUG() << "Encoding pauli constraints.";
+  splitXorR(initialPhase);
+  for (std::size_t q = 0U; q < N; ++q) {
+    const auto& xCol = vectorFromBitset(uncorrected.getBVFrom(q));
+    const auto& zCol = vectorFromBitset(uncorrected.getBVFrom(N + q));
+
+    LogicVector iChange{};
+    LogicVector xChange{};
+    LogicVector zChange{};
+    LogicVector yChange{};
+    iChange.reserve(S);
+    xChange.reserve(S);
+    zChange.reserve(S);
+    yChange.reserve(S);
+
+    for (std::size_t row = 0U; row < S; ++row) {
+      DEBUG() << "Encoding identity constraint.";
+      iChange.emplace_back(logicbase::LogicTerm::ite(
+          paulis[q][0], LogicTerm(false), LogicTerm(false)));
+      DEBUG() << "Encoding X constraint.";
+      xChange.emplace_back(
+          logicbase::LogicTerm::ite(paulis[q][1], zCol[row], LogicTerm(false)));
+      DEBUG() << "Encoding Y constraint.";
+      yChange.emplace_back(logicbase::LogicTerm::ite(
+          paulis[q][2], !(zCol[row] == xCol[row]), LogicTerm(false)));
+      DEBUG() << "Encoding Z constraint.";
+      zChange.emplace_back(
+          logicbase::LogicTerm::ite(paulis[q][3], xCol[row], LogicTerm(false)));
+    }
+    splitXorR(iChange);
+    splitXorR(xChange);
+    splitXorR(yChange);
+    splitXorR(zChange);
+
+    // at least one gate per qubit
+    lb->assertFormula(paulis[q][0] || paulis[q][1] || paulis[q][2] ||
+                      paulis[q][3]);
+  }
+  DEBUG() << "Assert Phase.";
+  for (std::size_t row = 0U; row < S; ++row) {
+    lb->assertFormula(targetPhase[row] == xorHelpers.back()[row]);
+  }
+  DEBUG() << "Finished Phase Encoding.";
+}
+
+void PhaseCorrectionEncoder::splitXorR(const LogicVector& changes) {
+  auto xorHelper = LogicVector{};
+  xorHelper.reserve(S);
+  for (std::size_t row = 0U; row < S; ++row) {
+    const std::string hName =
+        "h_" + std::to_string(row) + "_" + std::to_string(xorHelpers.size());
+    DEBUG() << "Creating helper variable for RChange XOR " << hName;
+    xorHelper.emplace_back(lb->makeVariable(hName));
+  }
+  xorHelpers.emplace_back(xorHelper);
+  if (xorHelpers.size() == 1) {
+    for (std::size_t row = 0U; row < S; ++row) {
+      lb->assertFormula(xorHelpers[0][row] == changes[row]);
+    }
+  } else {
+    for (std::size_t row = 0U; row < S; ++row) {
+      lb->assertFormula(
+          xorHelpers.back()[row] ==
+          (changes[row] != xorHelpers[xorHelpers.size() - 2][row]));
+    }
+  }
+}
```

Comment:
This method is already implemented elsewhere if I am not mistaken. Can we avoid the code duplication here?